modules/lang-painless/src/main/java/org/elasticsearch/painless/Whitelist.java

```diff
@@ -0,0 +1,802 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless;
+
+import java.io.InputStreamReader;
+import java.io.LineNumberReader;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+
+/**
+ * Whitelist contains data structures and methods designed to either load or generate a white-list of
+ * Java classes, methods, and fields that can be used within a Painless script.  Whitelist is used directly
+ * to create a new {@link Definition} and is passed into {@link Definition#Definition(Whitelist)}.
+ * {@link Definition}'s are then used as part of each Painless script at both compile-time and
+ * run-time to determine what a script is allowed to execute.
+ *
+ * Several single letters are used as prefixes for many variables in this class.  They mean the following:
+ * p - Used as part of the Painless typing system
+ * j - Used as part of the Java typing system
+ * w = Used as part of the Whitelist typing system
+ *
+ * A Whitelist consists of two pieces.  The first is a {@link String} {@link #pDynamicTypeName}
+ * that represent the singularly allowed name of the dynamic type in Painless.  Should this value be
+ * {@code null} then only static typing will be enabled.  The second is a list of {@link WStruct}s
+ * that represent the equivalent of a Java class in Painless known as a struct.  These will be the types
+ * available for use in Painless scripts compiled using a {@link Definition} generated by this Whitelist.
+ */
+public final class Whitelist {
+
+    /**
+     * WStruct represents the equivalent of a Java class in Painless complete with super classes,
+     * constructors, methods, and fields.  In Painless a class is known as a struct primarily to avoid
+     * naming conflicts internally.
+     *
+     * Structs in Painless allow for arity overloading for constructors and methods.  Arity overloading
+     * means that multiple constructors are allowed for a single struct as long as they have a different
+     * number of parameter types, and multiples methods with the same name are allowed for a single struct
+     * as long as they have the same return type and a different number of parameter types.
+     *
+     * Structs are allowed to extend other structs, known as a super struct, in which they will have
+     * available all of the methods and fields from a super struct.
+     */
+    static class WStruct {
+
+        /** The Painless name of this struct which will also be the name of a type in a Painless script.  */
+        final String pTypeName;
```

Comment:
Can we use the full prefix? I know it is more verbose, but it is easier to read. eg painlessTypeName, etc